File: 2025/day08/playground.py
"""Day 8: Playground - Junction box circuit connection problem."""
import math
from itertools import combinations
from pathlib import Path


def parse_input(filename):
    """Parse junction box coordinates from input file."""
    filepath = Path(__file__).parent / filename
    boxes = []
    with open(filepath) as f:
        for line in f:
            line = line.strip()
            if line:
                x, y, z = map(int, line.split(','))
                boxes.append((x, y, z))
    return boxes


def calculate_distance(box1, box2):
    """Calculate Euclidean distance between two 3D points."""
    return math.sqrt(
        (box1[0] - box2[0]) ** 2 +
        (box1[1] - box2[1]) ** 2 +
        (box1[2] - box2[2]) ** 2
    )


class UnionFind:
    """Union-Find data structure for tracking connected components."""
    
    def __init__(self, n):
        self.parent = list(range(n))
        self.size = [1] * n
    
    def find(self, x):
        """Find root of x with path compression."""
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]
    
    def union(self, x, y):
        """Union two sets. Returns True if they were different sets."""
        root_x = self.find(x)
        root_y = self.find(y)
        
        if root_x == root_y:
            return False
        
        # Union by size
        if self.size[root_x] < self.size[root_y]:
            self.parent[root_x] = root_y
            self.size[root_y] += self.size[root_x]
        else:
            self.parent[root_y] = root_x
            self.size[root_x] += self.size[root_y]
        
        return True
    
    def get_component_sizes(self):
        """Get sizes of all connected components."""
        components = {}
        for i in range(len(self.parent)):
            root = self.find(i)
            if root not in components:
                components[root] = 0
            components[root] += 1
        return sorted(components.values(), reverse=True)


def solve_part1(filename, num_edges=1000):
    """
    Process the first num_edges shortest edges, connecting junction boxes
    that aren't already in the same circuit. Return the product of the
    three largest circuit sizes.
    
    Args:
        filename: Input file with junction box coordinates
        num_edges: Number of edges to process from sorted list
    
    Returns:
        Product of three largest circuit sizes
    """
    boxes = parse_input(filename)
    n = len(boxes)
    
    # Generate all edges with distances
    edges = []
    for i, j in combinations(range(n), 2):
        dist = calculate_distance(boxes[i], boxes[j])
        edges.append((dist, i, j))
    
    # Sort by distance
    edges.sort()
    
    # Process first num_edges edges using Union-Find
    uf = UnionFind(n)
    for edge_idx in range(min(num_edges, len(edges))):
        _, i, j = edges[edge_idx]
        uf.union(i, j)
    
    # Get circuit sizes and calculate result
    sizes = uf.get_component_sizes()
    return sizes[0] * sizes[1] * sizes[2]


def solve_part2(filename):
    """
    Connect junction boxes until all are in one circuit. Return the product
    of the X coordinates of the last two boxes connected.
    
    Args:
        filename: Input file with junction box coordinates
    
    Returns:
        Product of X coordinates of last connection
    """
    boxes = parse_input(filename)
    n = len(boxes)
    
    # Generate all edges with distances
    edges = []
    for i, j in combinations(range(n), 2):
        dist = calculate_distance(boxes[i], boxes[j])
        edges.append((dist, i, j))
    
    # Sort by distance
    edges.sort()
    
    # Process edges until we have one circuit
    uf = UnionFind(n)
    uf.num_components = n  # Track number of components
    last_connection = None
    
    for dist, i, j in edges:
        if uf.union(i, j):
            last_connection = (i, j)
            uf.num_components -= 1
            
            # Stop when all boxes are in one circuit
            if uf.num_components == 1:
                break
    
    # Get X coordinates and multiply
    box_i, box_j = last_connection
    x1 = boxes[box_i][0]
    x2 = boxes[box_j][0]
    return x1 * x2


def _generate_sorted_edges(boxes):
    """Helper to generate and sort all edges."""
    edges = []
    for i, j in combinations(range(len(boxes)), 2):
        dist = calculate_distance(boxes[i], boxes[j])
        edges.append((dist, i, j))
    edges.sort()
    return edges


if __name__ == "__main__":
    import sys
    
    if len(sys.argv) < 2:
        print("Usage: python playground.py <input_file> [num_edges] [-d|--debug] [--part2]")
        print("Example: python playground.py example.txt 10")
        print("Example: python playground.py input.txt 1000 --debug")
        print("Example: python playground.py input.txt --part2")
        sys.exit(1)
    
    filename = sys.argv[1]
    debug = "-d" in sys.argv or "--debug" in sys.argv
    part2 = "--part2" in sys.argv
    num_edges = 1000
    for arg in sys.argv[2:]:
        if arg not in ["-d", "--debug", "--part2"]:
            num_edges = int(arg)
            break
    
    try:
        boxes = parse_input(filename)
        n = len(boxes)
        edges = _generate_sorted_edges(boxes)
        
        if part2:
            # Part 2: Connect until one circuit
            uf = UnionFind(n)
            num_components = n
            last_connection = None
            
            for edge_idx, (dist, i, j) in enumerate(edges):
                if uf.union(i, j):
                    last_connection = (i, j, dist)
                    num_components -= 1
                    if num_components == 1:
                        break
            
            box_i, box_j, dist = last_connection
            x1, x2 = boxes[box_i][0], boxes[box_j][0]
            result = x1 * x2
            
            print(f"Last connection: box{box_i} {boxes[box_i]} <-> box{box_j} {boxes[box_j]}")
            print(f"X coordinates: {x1}, {x2}")
            print(f"Result: {x1} × {x2} = {result}")
        else:
            # Part 1: Process fixed number of edges
            uf = UnionFind(n)
            for edge_idx in range(min(num_edges, len(edges))):
                _, i, j = edges[edge_idx]
                uf.union(i, j)
            
            sizes = uf.get_component_sizes()
            result = sizes[0] * sizes[1] * sizes[2]
            
            print(f"Total boxes: {n}")
            print(f"Circuits remaining: {len(sizes)}")
            print(f"Three largest: {sizes[0]}, {sizes[1]}, {sizes[2]}")
            print(f"Result: {sizes[0]} × {sizes[1]} × {sizes[2]} = {result}")
            
    except FileNotFoundError:
        print(f"Error: {filename} not found")
        sys.exit(1)
    except Exception as e:
        print(f"Error: {e}")
        import traceback
        traceback.print_exc()
        sys.exit(1)

File: 2025/day08/playground.pl
% Day 8: Playground - Prolog solution for junction box connections
% Connect boxes using minimum spanning tree approach

:- use_module(library(lists)).

% Dynamic predicates for union-find structure
:- dynamic(parent/2).
:- dynamic(rank/2).

% Parse input file to get 3D coordinates
load_boxes(Filename, Boxes) :-
    open(Filename, read, Stream),
    read_lines(Stream, Lines),
    close(Stream),
    maplist(parse_coordinate, Lines, Boxes).

% Read all lines from stream
read_lines(Stream, Lines) :-
    read_line_to_codes(Stream, Codes),
    (   Codes == end_of_file
    ->  Lines = []
    ;   atom_codes(Line, Codes),
        Lines = [Line|RestLines],
        read_lines(Stream, RestLines)
    ).

% Parse coordinate line like "1,2,3"
parse_coordinate(Line, coord(X, Y, Z)) :-
    atomic_list_concat(Parts, ',', Line),
    Parts = [XAtom, YAtom, ZAtom],
    atom_number(XAtom, X),
    atom_number(YAtom, Y),
    atom_number(ZAtom, Z).

% Calculate Euclidean distance between two 3D points
distance(coord(X1, Y1, Z1), coord(X2, Y2, Z2), Dist) :-
    DX is X2 - X1,
    DY is Y2 - Y1,
    DZ is Z2 - Z1,
    Dist is sqrt(DX*DX + DY*DY + DZ*DZ).

% Generate all pairs of boxes with distances
generate_edges(Boxes, Edges) :-
    length(Boxes, N),
    N1 is N - 1,
    numlist(0, N1, Indices),
    findall(edge(Dist, I, J), 
            (member(I, Indices), 
             member(J, Indices), 
             I < J,
             nth0(I, Boxes, BoxI),
             nth0(J, Boxes, BoxJ),
             distance(BoxI, BoxJ, Dist)), 
            Edges).

% Initialize Union-Find structure
init_union_find(N) :-
    retractall(parent(_, _)),
    retractall(rank(_, _)),
    N1 is N - 1,
    numlist(0, N1, Indices),
    maplist(init_node, Indices).

init_node(I) :-
    assertz(parent(I, I)),
    assertz(rank(I, 0)).

% Find root with path compression
find_root(X, Root) :-
    parent(X, P),
    (   X == P
    ->  Root = X
    ;   find_root(P, Root),
        retract(parent(X, P)),
        assertz(parent(X, Root))
    ).

% Union two sets by rank
union_sets(X, Y) :-
    find_root(X, RootX),
    find_root(Y, RootY),
    (   RootX == RootY
    ->  fail  % Already in same set
    ;   rank(RootX, RankX),
        rank(RootY, RankY),
        (   RankX < RankY
        ->  retract(parent(RootX, RootX)),
            assertz(parent(RootX, RootY))
        ;   RankX > RankY
        ->  retract(parent(RootY, RootY)),
            assertz(parent(RootY, RootX))
        ;   % RankX == RankY
            retract(parent(RootY, RootY)),
            assertz(parent(RootY, RootX)),
            retract(rank(RootX, RankX)),
            NewRank is RankX + 1,
            assertz(rank(RootX, NewRank))
        )
    ).

% Count components in Union-Find structure
count_components(N, Count) :-
    N1 is N - 1,
    numlist(0, N1, Indices),
    findall(Root, (member(I, Indices), find_root(I, Root)), Roots),
    sort(Roots, UniqueRoots),
    length(UniqueRoots, Count).

% Get component sizes
get_component_sizes(N, Sizes) :-
    N1 is N - 1,
    numlist(0, N1, Indices),
    findall(Root, (member(I, Indices), find_root(I, Root)), Roots),
    msort(Roots, SortedRoots),
    group_consecutive(SortedRoots, Groups),
    maplist(length, Groups, Sizes).

% Group consecutive identical elements
group_consecutive([], []).
group_consecutive([H|T], [[H|Group]|RestGroups]) :-
    take_while_equal(H, T, Group, Rest),
    group_consecutive(Rest, RestGroups).

take_while_equal(_, [], [], []).
take_while_equal(X, [X|T], [X|Group], Rest) :-
    take_while_equal(X, T, Group, Rest).
take_while_equal(X, [Y|T], [], [Y|T]) :-
    X \== Y.

% Part 1: Connect 1000 edges and find product of 3 largest components
part1(Filename, NumEdges, Result) :-
    load_boxes(Filename, Boxes),
    length(Boxes, N),
    generate_edges(Boxes, UnsortedEdges),
    sort(UnsortedEdges, SortedEdges),
    
    init_union_find(N),
    connect_n_edges(SortedEdges, NumEdges, 0),
    
    get_component_sizes(N, AllSizes),
    sort(0, @>=, AllSizes, SortedSizes),  % Sort descending
    SortedSizes = [Size1, Size2, Size3|_],
    Result is Size1 * Size2 * Size3.

% Connect N edges using Union-Find
connect_n_edges(_, 0, _) :- !.
connect_n_edges([edge(_, I, J)|RestEdges], N, Connected) :-
    (   union_sets(I, J)
    ->  N1 is N - 1,
        Connected1 is Connected + 1
    ;   N1 = N,
        Connected1 = Connected
    ),
    connect_n_edges(RestEdges, N1, Connected1).
connect_n_edges([_|RestEdges], N, Connected) :-
    connect_n_edges(RestEdges, N, Connected).

% Part 2: Connect until one component, return product of X coordinates of last connection
part2(Filename, Result) :-
    load_boxes(Filename, Boxes),
    length(Boxes, N),
    generate_edges(Boxes, UnsortedEdges),
    sort(UnsortedEdges, SortedEdges),
    
    init_union_find(N),
    connect_until_one(SortedEdges, Boxes, N, Result).

% Connect edges until only one component remains
connect_until_one([edge(_, I, J)|RestEdges], Boxes, N, Result) :-
    (   union_sets(I, J)
    ->  count_components(N, Count),
        (   Count == 1
        ->  % Found the last connection
            nth0(I, Boxes, coord(XI, _, _)),
            nth0(J, Boxes, coord(XJ, _, _)),
            Result is XI * XJ
        ;   connect_until_one(RestEdges, Boxes, N, Result)
        )
    ;   connect_until_one(RestEdges, Boxes, N, Result)
    ).

% Test predicates
test_example :-
    part1('example.txt', 10, Result),
    format('Example Part 1 (10 edges): ~w~n', [Result]).

test_main_part1 :-
    part1('input.txt', 1000, Result),
    format('Part 1 (1000 edges): ~w~n', [Result]).

test_main_part2 :-
    part2('input.txt', Result),
    format('Part 2: ~w~n', [Result]).

% Debug: Show first few edges
show_edges(Filename, Count) :-
    load_boxes(Filename, Boxes),
    generate_edges(Boxes, UnsortedEdges),
    sort(UnsortedEdges, SortedEdges),
    take(Count, SortedEdges, FirstEdges),
    forall(member(edge(Dist, I, J), FirstEdges),
           format('Edge ~w-~w: distance ~w~n', [I, J, Dist])).

% Utility: take first N elements
take(0, _, []) :- !.
take(_, [], []) :- !.
take(N, [H|T], [H|Result]) :-
    N > 0,
    N1 is N - 1,
    take(N1, T, Result).